lib/jsonapi/serializer.ex

```diff
@@ -233,26 +244,48 @@ defmodule JSONAPI.Serializer do
 
   defp merge_related_links(
          encoded_data,
-         {rel_view, rel_data, rel_url, conn},
+         {rel_view, _rel_key, rel_data, rel_url, conn},
          false = _remove_links
        ) do
-    Map.merge(encoded_data, %{links: %{self: rel_url, related: rel_view.url_for(rel_data, conn)}})
+    rel_data =
+      if id = encoded_data[:data][:id] do
+        %{id: id}
+      else
+        rel_data
+      end
+
+    if rel_data do
+      Map.merge(encoded_data, %{
+        links: %{self: rel_url, related: rel_view.url_for(rel_data, conn)}
+      })
+    else
+      encoded_data
+    end
   end
 
   defp merge_related_links(encoded_rel_data, _info, _remove_links), do: encoded_rel_data
 
-  @spec encode_rel_data(module(), map() | list()) :: map() | nil
-  def encode_rel_data(_view, nil), do: nil
+  @spec encode_rel_data(map(), String.t(), module(), map() | list()) :: map() | nil
+  def encode_rel_data(_entity, _key, _view, nil), do: nil
 
-  def encode_rel_data(view, data) when is_list(data) do
-    Enum.map(data, &encode_rel_data(view, &1))
+  def encode_rel_data(entity, key, view, data) when is_list(data) do
+    Enum.map(data, &encode_rel_data(entity, key, view, &1))
   end
 
-  def encode_rel_data(view, data) do
-    %{
-      type: view.type(),
-      id: view.id(data)
-    }
+  def encode_rel_data(entity, key, view, data) do
+    id =
+      if assoc_loaded?(data) do
+        view.id(data)
+      else
+        Map.get(entity, :"#{key}_id")
+      end
```

Comment:
Annotation: This is the main change.  Instead of just loading the ID if the `data` is loaded, get the ID from the foreign key if the data is not loaded.

Discussion Point: Line 280 could be improved by using Ecto to get the exact name of the foreign key:
```elixir
# Old
Map.get(entity, :"#{key}_id")

# New
%{owner_key: owner_key} = entity.__struct__.__schema__(:association, key)
```

The reason I did not do this is because it broke all of the Serializer tests.  Those tests are using a plain map as their input test data, but in reality I think this would always be a struct backed by an Ecto Schema.  If we want to make that change, we will need to include Ecto as a test dependency, and that was starting to make this PR too large.